lib/pam/pam.go

```diff
@@ -0,0 +1,326 @@
+// +build pam,cgo
+
+package pam
+
+// #cgo LDFLAGS: -ldl
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include <string.h>
+// #include <unistd.h>
+// #include <dlfcn.h>
+// #include <security/pam_appl.h>
+// extern char *library_name();
+// extern char* readCallback(int, int);
+// extern void writeCallback(int n, int s, char* c);
+// extern struct pam_conv *make_pam_conv(int);
+// extern int _pam_start(void *, const char *, const char *, const struct pam_conv *, pam_handle_t **);
+// extern int _pam_end(void *, pam_handle_t *, int);
+// extern int _pam_authenticate(void *, pam_handle_t *, int);
+// extern int _pam_acct_mgmt(void *, pam_handle_t *, int);
+// extern int _pam_open_session(void *, pam_handle_t *, int);
+// extern int _pam_close_session(void *, pam_handle_t *, int);
+// extern const char *_pam_strerror(void *, pam_handle_t *, int);
+import "C"
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentPAM,
+})
+
+// handler is used to register and find instances of *PAM at the package level
+// to enable callbacks from C code.
+type handler interface {
+	// writeStream will write to the output stream (stdout or stderr or
+	// equivlient).
+	writeStream(int, string) (int, error)
+
+	// readStream will read from the input stream (stdin or equivlient).
+	readStream(bool) (string, error)
+}
+
+var handlerMu sync.Mutex
+var handlerCount int
+var handlers map[int]handler = make(map[int]handler)
+
+//export writeCallback
+func writeCallback(index C.int, stream C.int, s *C.char) {
+	handlerMu.Lock()
```

Comment:
this will block all callbacks until this one's write will return, we should fix this